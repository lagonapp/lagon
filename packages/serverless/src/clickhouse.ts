import { workerData } from 'node:worker_threads';
import { ClickHouse } from 'clickhouse';

const clickhouse = new ClickHouse({});

workerData.forEach((functionResult, [functionId, deploymentId]) => {
  clickhouse
    .query(
      `INSERT INTO functions_result (date, functionId, deploymentId, cpuTime, memory, sendBytes, receivedBytes, requests) VALUES (now(), '${functionId}', '${deploymentId}', ${
        functionResult.reduce((acc, current) => {
          return acc + current.cpuTime;
        }, BigInt(0)) / BigInt(functionResult.length)
      }, ${
        functionResult.reduce((acc, current) => {
          return acc + current.memory;
        }, 0) / functionResult.length
      }, ${
        functionResult.reduce((acc, current) => {
          return acc + current.sentBytes;
        }, 0) / functionResult.length
      }, ${
        functionResult.reduce((acc, current) => {
          return acc + current.receivedBytes;
        }, 0) / functionResult.length
      }, ${functionResult.length})`,
    )
    .toPromise();

  functionResult.forEach(functionResult => {
    functionResult.logs.forEach(log => {
      clickhouse
        .query(
          `INSERT INTO logs (date, functionId, deploymentId, level, message) VALUES (now(), '${functionId}', '${deploymentId}', '${log.level}', '${log.content}')`,
        )
        .toPromise();
    });
  });
});
