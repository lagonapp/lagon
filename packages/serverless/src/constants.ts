import { cpus } from 'node:os';

export const IS_DEV = process.env.NODE_ENV !== 'production';
export const CPUS = cpus().length;
