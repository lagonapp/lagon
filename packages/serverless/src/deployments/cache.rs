use std::{
    collections::HashMap,
    sync::Arc,
    time::{Duration, Instant},
};

use log::info;
use tokio::sync::RwLock;

use crate::ISOLATES;

const CACHE_TASK_INTERVAL: Duration = Duration::from_secs(60);

pub fn run_cache_clear_task(last_requests: Arc<RwLock<HashMap<String, Instant>>>) {
    let isolates_cache_seconds = Duration::from_secs(
        dotenv::var("LAGON_ISOLATES_CACHE_SECONDS")
            .expect("LAGON_ISOLATES_CACHE_SECONDS must be set")
            .parse()
            .expect("Failed to parse LAGON_ISOLATES_CACHE_SECONDS"),
    );

    tokio::spawn(async move {
        loop {
            tokio::time::sleep(CACHE_TASK_INTERVAL).await;

            let now = Instant::now();
            let mut isolates_to_clear = Vec::new();
            let last_requests_reader = last_requests.read().await;

            info!("Running cache clear task");

            for (hostname, last_request) in last_requests_reader.iter() {
                if now.duration_since(*last_request) > isolates_cache_seconds {
                    isolates_to_clear.push(hostname.clone());
                }
            }

            // drop(last_requests);

            if isolates_to_clear.is_empty() {
                continue;
            }

            // Drop the read lock because we now acquire a write lock
            drop(last_requests_reader);

            let mut thread_isolates = ISOLATES.write().await;
            let mut last_requests = last_requests.write().await;

            for hostname in isolates_to_clear {
                for isolates in thread_isolates.values_mut() {
                    last_requests.remove(&hostname);

                    if let Some(isolate) = isolates.remove(&hostname) {
                        let (deployment, ..) = isolate
                            .get_metadata()
                            .unwrap_or_else(|| ("Unknown".to_owned(), "".to_owned()));

                        info!(deployment = deployment; "Clearing deployment from cache due to expiration");
                    }
                }
            }
        }
    });
}
