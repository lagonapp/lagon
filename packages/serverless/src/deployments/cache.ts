import { Deployment } from '@lagon/runtime';

const LAGON_REGION = process.env.LAGON_REGION;

if (!LAGON_REGION) {
  throw new Error('LAGON_REGION is not set');
}

const deployments = new Map<string, Deployment>();

export function getDeployments(): Map<string, Deployment> {
  return deployments;
}

export function setDeployment(domain: string, deployment: Deployment) {
  deployments.set(domain, {
    ...deployment,
    env: {
      ...deployment.env,
      LAGON_REGION: LAGON_REGION as string,
    },
  });
}

export function deleteDeployment(domain: string) {
  deployments.delete(domain);
}

export function clearDeployments() {
  deployments.clear();
}
