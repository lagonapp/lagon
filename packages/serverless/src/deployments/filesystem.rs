use std::io::Write;
use std::path::PathBuf;
use std::{env, fs, io, path::Path};

use log::info;

use super::Deployment;

pub fn create_deployments_folder() -> io::Result<()> {
    let path = Path::new("deployments");

    if !path.exists() {
        fs::create_dir(path)?;
        info!("Created deployments folder");
    }

    Ok(())
}

pub fn has_deployment_code(deployment: &Deployment) -> bool {
    let path = Path::new("deployments").join(deployment.id.clone() + ".js");

    path.exists()
}

pub fn get_deployment_code(deployment: &Deployment) -> io::Result<String> {
    let path = Path::new(env::current_dir().unwrap().as_path())
        .join("deployments")
        .join(deployment.id.clone() + ".js");
    let code = fs::read_to_string(path)?;

    Ok(code)
}

pub fn write_deployment(deployment_id: &str, buf: &[u8]) -> io::Result<()> {
    let mut file =
        fs::File::create(Path::new("deployments").join(deployment_id.to_owned() + ".js"))?;

    file.write_all(buf)?;
    info!("Wrote deployment: {}", deployment_id);

    Ok(())
}

pub fn write_deployment_asset(deployment_id: &str, asset: &str, buf: &[u8]) -> io::Result<()> {
    let asset = asset.replace("public/", "");
    let asset = asset.as_str();

    let dir = PathBuf::from("deployments")
        .join(deployment_id)
        .join(PathBuf::from(asset).parent().unwrap());
    fs::create_dir_all(dir)?;

    let mut file =
        fs::File::create(Path::new("deployments").join(deployment_id.to_owned() + "/" + asset))?;

    file.write_all(buf)?;
    info!("Wrote deployment ({}) asset: {}", deployment_id, asset);

    Ok(())
}

pub fn rm_deployment(deployment_id: &str) -> io::Result<()> {
    fs::remove_file(Path::new("deployments").join(deployment_id.to_owned() + ".js"))?;

    // It's possible that the folder doesn't exists if the deployment has no assets
    fs::remove_dir(Path::new("deployments").join(deployment_id.clone())).unwrap_or(());
    info!("Deleted deployment: {}", deployment_id);

    Ok(())
}
