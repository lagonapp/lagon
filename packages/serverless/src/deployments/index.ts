import path from 'node:path';
import fs from 'node:fs';
import { Deployment } from '@lagon/runtime';

const DEPLOYMENTS_FOLDER = path.join(path.resolve(), 'dist', 'deployments');

export function hasDeploymentCodeLocally(deployment: Deployment): boolean {
  const file = path.join(DEPLOYMENTS_FOLDER, `${deployment.deploymentId}.js`);

  return fs.existsSync(file);
}

export async function getDeploymentCode(deployment: Deployment): Promise<string> {
  const file = path.join(DEPLOYMENTS_FOLDER, `${deployment.deploymentId}.js`);

  return fs.readFileSync(file, 'utf8');
}

export function writeDeploymentCode(deployment: Deployment, code: string): void {
  const file = path.join(DEPLOYMENTS_FOLDER, `${deployment.deploymentId}.js`);

  if (deployment.assets.length > 0) {
    fs.mkdirSync(path.join(DEPLOYMENTS_FOLDER, deployment.deploymentId));
  }

  fs.writeFileSync(file, code, 'utf8');
}

export function getAssetContent(deployment: Deployment, name: string): string {
  const file = path.join(DEPLOYMENTS_FOLDER, deployment.deploymentId, name);

  return fs.readFileSync(file, 'utf8');
}

export function writeAssetContent(name: string, content: string): void {
  const file = path.join(DEPLOYMENTS_FOLDER, name);

  fs.writeFileSync(file, content, 'utf8');
}

export function deleteDeploymentCode(deployment: Deployment) {
  const file = path.join(DEPLOYMENTS_FOLDER, `${deployment.deploymentId}.js`);

  try {
    fs.rmSync(file);
    // Folder for assets
    fs.rmSync(path.join(DEPLOYMENTS_FOLDER, deployment.deploymentId), { recursive: true, force: true });
  } catch (e) {
    console.error(e);
  }
}
