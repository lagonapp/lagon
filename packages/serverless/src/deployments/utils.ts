import { Deployment } from '@lagon/runtime';
import { FastifyReply, FastifyRequest } from 'fastify';

export const deployments = new Map<string, Deployment>();

export function getDeploymentFromRequest(req: FastifyRequest): Deployment | undefined {
  const { host } = req.headers;

  if (!host) {
    return;
  }

  return deployments.get(host);
}

function getBytesFromHeaders(headers: { [key: string]: string | string[] | undefined }): number {
  return Object.entries(headers).reduce((acc, [key, value]) => {
    const items = [key];

    if (typeof value === 'object') {
      items.concat(value);
    } else if (value) {
      items.push(value);
    }

    acc += Buffer.byteLength(items.join(''));

    return acc;
  }, 0);
}

export function getBytesFromRequest(request: FastifyRequest): number {
  return Number(request.headers['content-length'] || 0) + getBytesFromHeaders(request.headers);
}

export function getBytesFromReply(reply: FastifyReply): number {
  return (
    Number(reply.getHeader('content-length') || 0) +
    getBytesFromHeaders(reply.headers as unknown as { [key: string]: string | string[] | undefined })
  );
}
