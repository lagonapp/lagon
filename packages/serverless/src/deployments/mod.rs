use std::{
    collections::{HashMap, HashSet},
    fs,
    path::Path,
    sync::Arc,
};

use anyhow::{anyhow, Result};
use log::{error, info};
use mysql::{prelude::Queryable, PooledConn};
use s3::Bucket;
use tokio::sync::RwLock;

use crate::deployments::filesystem::has_deployment_code;

use self::filesystem::{
    create_deployments_folder, rm_deployment, write_deployment, write_deployment_asset,
};

pub mod assets;
pub mod cache;
pub mod filesystem;
pub mod pubsub;

#[derive(Debug, Clone)]
pub struct Deployment {
    pub id: String,
    pub function_id: String,
    pub function_name: String,
    pub domains: HashSet<String>,
    pub assets: HashSet<String>,
    pub environment_variables: HashMap<String, String>,
    pub memory: usize,          // in MB (MegaBytes)
    pub timeout: usize,         // in ms (MilliSeconds)
    pub startup_timeout: usize, // in ms (MilliSeconds)
    pub is_production: bool,
}

impl Deployment {
    pub fn get_domains(&self) -> Vec<String> {
        let mut domains = Vec::new();

        domains.push(format!(
            "{}.{}",
            self.id,
            dotenv::var("LAGON_ROOT_DOMAIN").expect("LAGON_ROOT_DOMAIN must be set")
        ));

        // Default domain (function's name) and custom domains are only set in production deployments
        if self.is_production {
            domains.push(format!(
                "{}.{}",
                self.function_name,
                dotenv::var("LAGON_ROOT_DOMAIN").expect("LAGON_ROOT_DOMAIN must be set")
            ));

            domains.extend(self.domains.clone());
        }

        domains
    }

    pub async fn download(&self, bucket: &Bucket) -> Result<()> {
        match bucket.get_object(self.id.clone() + ".js").await {
            Ok(object) => {
                write_deployment(&self.id, object.bytes())?;

                if !self.assets.is_empty() {
                    for asset in &self.assets {
                        match bucket
                            .get_object(self.id.clone() + "/" + asset.as_str())
                            .await
                        {
                            Ok(object) => write_deployment_asset(&self.id, asset, object.bytes())?,
                            Err(error) => return Err(anyhow!(error)),
                        };
                    }
                }

                Ok(())
            }
            Err(error) => Err(anyhow!(error)),
        }
    }
}

pub async fn get_deployments(
    mut conn: PooledConn,
    bucket: Bucket,
) -> Result<Arc<RwLock<HashMap<String, Arc<Deployment>>>>> {
    let deployments = Arc::new(RwLock::new(HashMap::new()));

    let mut deployments_list: HashMap<String, Deployment> = HashMap::new();

    conn.query_map(
        r"
        SELECT
            Deployment.id,
            Deployment.isProduction,
            Function.id,
            Function.name,
            Function.memory,
            Function.timeout,
            Function.startupTimeout,
            Domain.domain,
            Asset.name
        FROM
            Deployment
        INNER JOIN Function
            ON Deployment.functionId = Function.id
        LEFT JOIN Domain
            ON Function.id = Domain.functionId
        LEFT JOIN Asset
            ON Deployment.id = Asset.deploymentId
    ",
        |(
            id,
            is_production,
            function_id,
            function_name,
            memory,
            timeout,
            startup_timeout,
            domain,
            asset,
        ): (
            String,
            bool,
            String,
            String,
            usize,
            usize,
            usize,
            Option<String>,
            Option<String>,
        )| {
            deployments_list
                .entry(id.clone())
                .and_modify(|deployment| {
                    if let Some(domain) = domain.clone() {
                        deployment.domains.insert(domain);
                    }

                    if let Some(asset) = asset.clone() {
                        deployment.assets.insert(asset);
                    }
                })
                .or_insert(Deployment {
                    id,
                    function_id,
                    function_name,
                    domains: domain
                        .map(|domain| {
                            let mut domains = HashSet::new();
                            domains.insert(domain);
                            domains
                        })
                        .unwrap_or_default(),
                    assets: asset
                        .map(|asset| {
                            let mut assets = HashSet::new();
                            assets.insert(asset);
                            assets
                        })
                        .unwrap_or_default(),
                    environment_variables: HashMap::new(),
                    memory,
                    timeout,
                    startup_timeout,
                    is_production,
                });
        },
    )?;

    let deployments_list: Vec<Deployment> = deployments_list.values().cloned().collect();

    info!("Found {} deployment(s) to deploy", deployments_list.len());

    if let Err(error) = create_deployments_folder() {
        error!("Could not create deployments folder: {}", error);
    }

    if let Err(error) = delete_old_deployments(&deployments_list).await {
        error!("Failed to delete old deployments: {:?}", error);
    }

    {
        let mut deployments = deployments.write().await;

        for deployment in deployments_list {
            if !has_deployment_code(&deployment) {
                if let Err(error) = deployment.download(&bucket).await {
                    error!("Failed to download deployment {}: {}", deployment.id, error);
                    continue;
                }
            }

            for domain in deployment.get_domains() {
                deployments.insert(domain, Arc::new(deployment.clone()));
            }
        }
    }

    Ok(deployments)
}

async fn delete_old_deployments(deployments: &[Deployment]) -> Result<()> {
    info!("Deleting old deployments");
    let local_deployments_files = fs::read_dir(Path::new("deployments"))?;

    for local_deployment_file in local_deployments_files {
        let local_deployment_file_name = local_deployment_file?
            .file_name()
            .into_string()
            .unwrap_or_else(|_| "".into());

        // Skip folders
        if !local_deployment_file_name.ends_with(".js") {
            continue;
        }

        let local_deployment_id = local_deployment_file_name.replace(".js", "");

        if !deployments
            .iter()
            .any(|deployment| deployment.id == local_deployment_id)
        {
            rm_deployment(&local_deployment_id)?;
        }
    }
    info!("Old deployments deleted");

    Ok(())
}
