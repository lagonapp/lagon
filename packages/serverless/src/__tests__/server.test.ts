import { Deployment, fetch } from '@lagon/runtime';
import startServer from 'src/server';
import { beforeAll, describe, expect, it, vi } from 'vitest';
import * as deploymentsConfig from 'src/deployments/utils';
import * as deployments from 'src/deployments';

const getDeployment = (): Deployment => ({
  functionId: 'functionId',
  functionName: 'functionName',
  deploymentId: Math.random().toString(),
  domains: [],
  memory: 128,
  timeout: 50,
  env: {},
  isCurrent: false,
  assets: [],
});

const request = async (code: string, init?: Parameters<typeof fetch>[1]): ReturnType<typeof fetch> => {
  const deployment = getDeployment();
  vi.spyOn(deploymentsConfig, 'getDeploymentFromRequest').mockReturnValue(deployment);
  vi.spyOn(deployments, 'getDeploymentCode').mockResolvedValue(code);

  const response = await fetch('http://localhost:12346/', init);
  if (response.options.headers && 'date' in response.options.headers) {
    delete response.options.headers.date;
  }

  return response;
};

beforeAll(async () => {
  startServer(12346, 'localhost');

  // Wait for the server to start
  await new Promise(resolve => setTimeout(resolve, 500));
});

describe('Server', () => {
  it('should reply text', async () => {
    const response = await request(`export function handler(request) {
  return new Response('Hello World');
}`);

    expect(response).toMatchSnapshot();
  });

  it('should reply json', async () => {
    const response = await request(`export function handler(request) {
  return new Response(JSON.stringify({ hello: 'world' }), {
    headers: {
      'Content-Type': 'application/json',
    }
  });
}`);

    expect(response).toMatchSnapshot();
  });

  it('should reply custom headers', async () => {
    const response = await request(`export function handler(request) {
  return new Response("Hello World", {
    headers: {
      'x-custom-header': 'custom-value',
    }
  });
}`);

    expect(response).toMatchSnapshot();
  });

  it('should reply custom status', async () => {
    const response = await request(`export function handler(request) {
  return new Response("Not Found", {
    status: 404,
    statusText: "Not Found",
  });
}`);

    expect(response).toMatchSnapshot();
  });

  it('should reply Uint8Array', async () => {
    const response = await request(`export function handler(request) {
  const body = new TextEncoder().encode('Hello World');
  return new Response(body);
}`);

    expect(response).toMatchSnapshot();
  });

  it('should handle json request', async () => {
    const response = await request(
      `export async function handler(request) {
  const body = await request.json();
  return new Response(body, {
    headers: {
      'Content-Type': 'application/json',
    }
  });
}`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({ hello: 'world' }),
      },
    );

    expect(response).toMatchSnapshot();
  });
});
