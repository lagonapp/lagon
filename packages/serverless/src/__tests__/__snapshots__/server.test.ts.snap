// Vitest Snapshot v1

exports[`Server > should handle json request 1`] = `
{
  "body": "{\\"hello\\":\\"world\\"}",
  "options": {
    "headers": {
      "connection": "close",
      "content-length": "17",
      "content-type": "application/json; charset=utf-8",
    },
    "status": 200,
    "statusText": "OK",
    "url": "",
  },
}
`;

exports[`Server > should reply custom headers 1`] = `
{
  "body": "Hello World",
  "options": {
    "headers": {
      "connection": "close",
      "content-length": "11",
      "content-type": "text/plain; charset=utf-8",
      "x-custom-header": "custom-value",
    },
    "status": 200,
    "statusText": "OK",
    "url": "",
  },
}
`;

exports[`Server > should reply custom status 1`] = `
{
  "body": "Not Found",
  "options": {
    "headers": {
      "connection": "close",
      "content-length": "9",
      "content-type": "text/plain; charset=utf-8",
    },
    "status": 404,
    "statusText": "Not Found",
    "url": "",
  },
}
`;

exports[`Server > should reply json 1`] = `
{
  "body": "{\\"hello\\":\\"world\\"}",
  "options": {
    "headers": {
      "connection": "close",
      "content-length": "17",
      "content-type": "application/json; charset=utf-8",
    },
    "status": 200,
    "statusText": "OK",
    "url": "",
  },
}
`;

exports[`Server > should reply text 1`] = `
{
  "body": "Hello World",
  "options": {
    "headers": {
      "connection": "close",
      "content-length": "11",
      "content-type": "text/plain; charset=utf-8",
    },
    "status": 200,
    "statusText": "OK",
    "url": "",
  },
}
`;
